File: app/src/main/java/com/mc0239/recyclertableviewexample/samples/FragmentSampleInteractive.java
package com.mc0239.recyclertableviewexample.samples;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.mc0239.recyclertableview.RecyclerTableView;
import com.mc0239.recyclertableview.RecyclerTableViewAdapter;
import com.mc0239.recyclertableviewexample.R;
import com.mc0239.recyclertableviewexample.SampleUserGenerator;
import com.mc0239.recyclertableviewexample.rows.UserCheckable;

import java.util.ArrayList;
import java.util.List;

public class FragmentSampleInteractive extends Fragment {

    private RecyclerTableView recyclerTableView;
    private RecyclerTableViewAdapter<UserCheckable> recyclerTableViewAdapter;

    private Button buttonAdd;
    private Button buttonAddAt;
    private Button buttonAddManyAt;
    private Button buttonRemove;

    private UserCheckable lastUser;
    private List<UserCheckable> lastUsers;

    public FragmentSampleInteractive() {
        // Prepare some data for the table
        ArrayList<UserCheckable> users = new ArrayList<>();
        for(int i = 0; i < 10; i++) {
            UserCheckable u = SampleUserGenerator.generateUser();
            users.add(u);
        }

        // Create a new RecyclerTableViewAdapter
        recyclerTableViewAdapter = new RecyclerTableViewAdapter<>(UserCheckable.class, users);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_sample_interactive, container, false);

        // Get RecyclerTableView view
        recyclerTableView = view.findViewById(R.id.recyclerTableView1);

        // Set adapter to RecyclerTableView
        recyclerTableView.setAdapter(recyclerTableViewAdapter);

        view.findViewById(R.id.buttonAdd).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                lastUser = SampleUserGenerator.generateUser();
                recyclerTableViewAdapter.addItem(lastUser);
            }
        });

        view.findViewById(R.id.buttonAddAt).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                lastUser = SampleUserGenerator.generateUser();
                recyclerTableViewAdapter.addItemAt(5, lastUser);
            }
        });

        view.findViewById(R.id.buttonAddMore).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                lastUsers = new ArrayList<>(3);
                lastUsers.add(SampleUserGenerator.generateUser());
                lastUsers.add(SampleUserGenerator.generateUser());
                lastUsers.add(SampleUserGenerator.generateUser());
                recyclerTableViewAdapter.addItems(lastUsers);
            }
        });

        view.findViewById(R.id.buttonAddMoreAt).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                lastUsers = new ArrayList<>(3);
                lastUsers.add(SampleUserGenerator.generateUser());
                lastUsers.add(SampleUserGenerator.generateUser());
                lastUsers.add(SampleUserGenerator.generateUser());
                recyclerTableViewAdapter.addItemsAt(5, lastUsers);
            }
        });

        view.findViewById(R.id.buttonRemove).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (lastUser != null)
                    recyclerTableViewAdapter.removeItem(lastUser);
                lastUser = null;
            }
        });

        view.findViewById(R.id.buttonRemoveAt).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (recyclerTableViewAdapter.getItemCount() > 5)
                    recyclerTableViewAdapter.removeItemAt(5);
            }
        });

        view.findViewById(R.id.buttonRemoveMore).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (lastUsers != null)
                    recyclerTableViewAdapter.removeItems(lastUsers);
                lastUsers = null;
            }
        });

        view.findViewById(R.id.buttonClear).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                recyclerTableViewAdapter.clearItems();
            }
        });

        return view;
    }

}

File: app/src/main/kotlin/com/mc0239/recyclertableviewexample/samples/FragmentSampleInteractive.kt
package com.mc0239.recyclertableviewexample.samples

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import com.mc0239.recyclertableview.RecyclerTableViewAdapter
import com.mc0239.recyclertableviewexample.databinding.FragmentSampleInteractiveBinding
import com.mc0239.recyclertableviewexample.rows.UserCheckable
import com.mc0239.recyclertableviewexample.util.SampleUserGenerator

class FragmentSampleInteractive : Fragment() {

    private lateinit var binding: FragmentSampleInteractiveBinding

    private val recyclerTableViewAdapter: RecyclerTableViewAdapter<UserCheckable>

    init {
        // Prepare some data for the table
        val users = mutableListOf<UserCheckable>()
        // Stress-test: generate 10 000 entries!!
        for (i in 0 until 10_000) {
            users.add(SampleUserGenerator.generateUser())
        }

        recyclerTableViewAdapter = RecyclerTableViewAdapter(UserCheckable::class.java, users)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentSampleInteractiveBinding.inflate(inflater, container, false)

        binding.recyclerTableView1.setAdapter(recyclerTableViewAdapter)

        binding.buttonAdd.setOnClickListener {
            recyclerTableViewAdapter.addItem(SampleUserGenerator.generateUser())
        }

        binding.buttonAddAt.setOnClickListener {
            recyclerTableViewAdapter.addItemAt(minOf(recyclerTableViewAdapter.itemCount, 5), SampleUserGenerator.generateUser())
        }

        binding.buttonAddMore.setOnClickListener {
            recyclerTableViewAdapter.addItems((1 until 100).map { SampleUserGenerator.generateUser() })
        }

        binding.buttonAddMoreAt.setOnClickListener {
            recyclerTableViewAdapter.addItemsAt(minOf(recyclerTableViewAdapter.itemCount, 5), (1 until 100).map { SampleUserGenerator.generateUser() })
        }

        binding.buttonRemove.setOnClickListener {
            if (recyclerTableViewAdapter.itemCount > 0) {
                recyclerTableViewAdapter.removeItemAt(0)
            }
        }

        binding.buttonRemoveAt.setOnClickListener {
            if (recyclerTableViewAdapter.itemCount >= 5) {
                recyclerTableViewAdapter.removeItemAt(5)
            }
        }

        binding.buttonRemoveMore.setOnClickListener {
            val items = recyclerTableViewAdapter.items.slice(0 until minOf(recyclerTableViewAdapter.itemCount, 100))
            recyclerTableViewAdapter.removeItems(items)
        }

        binding.buttonRemoveChecked.setOnClickListener {
            recyclerTableViewAdapter.removeItems(recyclerTableViewAdapter.checkedItems)
        }

        binding.buttonClear.setOnClickListener {
            recyclerTableViewAdapter.clearItems()
        }

        binding.buttonCheckEllies.setOnClickListener {
            for (u in recyclerTableViewAdapter.items) {
                u.checked = u.name == "Ellie"
            }

            // Important: If doing any changes with the objects in rows, do not forget to notify
            // the adapter that there were changes made!
            recyclerTableViewAdapter.notifyDataSetChanged()
        }

        return binding.root
    }

}